src/installer/corehost/common/pal.windows.cpp

```diff
@@ -96,11 +96,11 @@ void* pal::map_file_readonly(const pal::string_t& path, size_t &length)
     length = (size_t)fileSize.QuadPart;
 
     HANDLE map = CreateFileMappingW(file, NULL, PAGE_READONLY, 0, 0, NULL);
+    CloseHandle(file);
 
     if (map == NULL)
     {
         trace::warning(_X("Failed to map file. CreateFileMappingW(%s) failed with error %d"), path.c_str(), GetLastError());
-        CloseHandle(file);
         return nullptr;
     }
 
```

Comment:
Do you also need to close `map` handle before the function returns?

Reply:
There is no API to close a File Mapping. The [UnMapViewOfFile](https://docs.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-unmapviewoffile) documentation says:  
"Although an application may close the file handle used to create a file mapping object, the system holds the corresponding file open until the last view of the file is unmapped. Files for which the last view has not yet been unmapped are held open with no sharing restrictions." 

[CreateFileMapping](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createfilemappinga) documentation says: 
"Mapped views of a file mapping object maintain internal references to the object, and a file mapping object does not close until all references to it are released. Therefore, to fully close a file mapping object, an application must unmap all mapped views of the file mapping object by calling UnmapViewOfFile and close the file mapping object handle by calling CloseHandle. These functions can be called in any order."

Therefore, I think `map` handle cannot be closed; the sharing violation was because of the `FileHandle`.

Reply:
The `map` handle should be closed using `CloseHandle`. See https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createfilemappinga :

> Therefore, to fully close a file mapping object, an application must unmap all mapped views of the file mapping object by calling UnmapViewOfFile and close the file mapping object handle by calling CloseHandle. These functions can be called in any order.

You should close it even if it does not cause any obvious problem like sharing violation. It is good practice to avoid leaking resources.

Reply:
Thanks @jkotas. I misunderstood the text to mean that the mapping handle is automatically closed after all the references are unmapped. I've fixed it now.